libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/DefaultTracer.java

```diff
@@ -27,31 +27,42 @@ class DefaultTracer implements Tracer {
     static final String THREAD_NAME = "th_name";
 
     private final TracingTelemetry tracingTelemetry;
-    private final TracerContextStorage<String, Span> tracerContextStorage;
+    private final TracerContextStorage<String, Span> spanTracerContextStorage;
```

Comment:
I believe the TracerContextStorage should only keep the `SpanScope`, not `Span`: as we discussed, the scope is responsible for managing context (like thread locals fe) and `Span` is not, it should not leak into context. Since `SpanScope` has `getSpan`, it is always possible to get the `Span` from the scope.

Reply:
@reta, Thanks for reviewing this. If we close SpanScope on one thread and still pass it to other thread through context propagation then I think lifecycle or context of the SpanScope would be little confusing.

Are we good to define the definition of SpanScope in a way that SpanScope::close means clearing the SpanScope from the current thread (like ThreadLocal you mentioned) and same SpanScope still may exist in the other threads. To end a Span we need to execute SpanScope::getSpan::endSpan and then SpanScope::close to clear up the context.

Reply:
Little complexity here would be to recreate the SpanScope objects as it may be containing the previous SpanScope's state(Parent of this SpanScope)

Reply:
> @reta, Thanks for reviewing this. If we close SpanScope on one thread and still pass it to other thread through context propagation then I think lifecycle or context of the SpanScope would be little confusing.

Oh I see, this is indeed the complexity because we stash and pass the context around, ideally the SpanScope should be only local to current thread. Shouldn't we just use `ThreadLocal` for scopes than?

Reply:
Yes, we can use the ThreadLocal (I also thought of the same initially). Since we are managing the ThreadLocal all over through ThreadContext then thought of using the same. If you can take a look at the `ThreadContextSpanScopeStorage` it's being used as a ThreadLocal only as there are no propagators of transient header.

Reply:
Updated the storage to ThreadLocal for SpanScope.